File: Difficulty: Medium/Opposite Sign Pair Reduction/opposite-sign-pair-reduction.cpp
class Solution {
	public:
	vector<int> reducePairs(vector<int>& arr) {
		stack<int> s;
		
		for (int x : arr) {
			
			while (!s.empty() &&
			((s.top() > 0 && x < 0) ||
			(s.top() < 0 && x > 0))) {
				
				if (abs(s.top()) > abs(x)) {
					x = s.top();
					s.pop();
				}
				else if (abs(s.top()) < abs(x)) {
					s.pop();
				}
				else {
					s.pop();
					x = 0;
					break;
				}
			}
			
			if (x != 0)
				s.push(x);
		}
		
		vector<int> ans(s.size());
		
		for (int i = ans.size() - 1; i >= 0; i--) {
			ans[i] = s.top();
			s.pop();
		}
		
		return ans;
	}
};

File: Difficulty: Medium/Opposite Sign Pair Reduction/opposite-sign-pair-reduction.py
from collections import deque
from typing import List
class Solution:
    
    def reducePairs(self, arr):
        # code here
        st=deque()
        
        for a in arr:
            if not st or (st[-1]<0 and a<0) or (st[-1]>0 and a>0):
                st.append(a)
            else:
                while(st and abs(st[-1])<abs(a)):
                    st.pop()
                
                if(st and abs(st[-1])==abs(a)):
                    st.pop()
                    continue
                if not st:
                    st.append(a)
        
        return list(st)
                    